src/traits.rs

```diff
@@ -0,0 +1,45 @@
+use Symbol;
+
+/// An object file.
+pub trait Object<'a>: Sized {
+    /// A section in the object file.
+    type Section: ObjectSection<'a>;
+
+    /// An iterator over the sections in the object file.
+    type SectionIterator: Iterator<Item = Self::Section>;
+
+    /// Parse the raw object file data.
+    fn parse(data: &'a [u8]) -> Result<Self, &'static str>;
+
+    /// Get the contents of the section named `section_name`, if such
+    /// a section exists.
+    fn get_section(&self, section_name: &str) -> Option<&'a [u8]>;
+
+    /// Get an iterator over the sections in the file.
+    // TODO: avoid 'a on self using Associated Type Constructor
+    fn get_sections(&'a self) -> Self::SectionIterator;
```

Comment:
Is the idea that the section iterator has the lifetime of the underlying data, not of `self`?

Reply:
So `'a` is normally the lifetime of the underlying data, but here we're further restricting it to the lifetime of self, which should be unnecessary. The section iterator needs to be bounded by the lifetime of self, but the data in the section should have the lifetime of the underlying data. So what I think we need is:

```
pub trait Object<'a> {
    type Section: ObjectSection<'a>;
    type SectionIterator<'me>: Iterator<Item = Self::Section>;
    fn get_sections(&'me self) -> Self::SectionIterator<'me>;
}
```

For the user, the difference this will make is that if they use `Object::get_sections()` followed by `ObjectSection::data()`, the section data that is returned will be valid after that object is dropped. I don't think this is something that will be commonly required, because it's not difficult to create the object at the same time as your memmap of the data.